syntax = "proto3";

package couponmanage;

service CouponManager {
  rpc RegisterCafe (CafeDetail) returns (CafeRegisterAndUpdateReply) {}
  rpc ReceiveAllCafes (EmptyRequest) returns (AllCafeReply) {}
  rpc ReceiveCafeDetail (CafeDetailRequest) returns (CafeDetail) {}
  rpc UpdateCafe (CafeUpdateRequest) returns (CafeRegisterAndUpdateReply) {}
}

enum DayOfTheWeek {
    SUN = 0;
    MON = 1;
    TUE = 2;
    WED = 3;
    THU = 4;
    FRI = 5;
    SAT = 6;
}

message OperatingInfo {
    DayOfTheWeek open_day_of_the_week = 1;
    string open_hour = 2;
    DayOfTheWeek close_day_of_the_week = 3;
    string close_hour = 4;
}

message CafeDetail {
  string name = 1;
  string address = 2;
  double rating = 3;
  repeated OperatingInfo operating_info = 4;
}

message CafeUpdateRequest {
  string name = 1;
  string address = 2;
  double rating = 3;
  repeated OperatingInfo operating_info = 4;
  bool is_address_null = 5;
  bool is_rating_null = 6;
  bool is_operating_info_null = 7;
}

message AllCafeReply {
  repeated CafeDetail cafes = 1;
}

message CafeDetailRequest {
  string name = 1;
}

message EmptyRequest {}
message EmptyReply {}

message CafeRegisterAndUpdateReply {
  bool is_success = 1;
}